File: typescript/src/adapters/inventory.ts
/**
 * In-memory inventory adapter with atomic reserve-all.
 */

import type { InventoryGateway, ReservationToken } from "../application/ports";
import { CompensationFailure, InsufficientStock } from "../domain/errors";
import type { OrderId, OrderLine } from "../domain/order";
import type { Sku } from "../domain/sku";

/**
 * `InventoryGateway` double enforcing finite stock, for tests and demos.
 *
 * The stock map is keyed by the branded `Sku`, so raw strings cannot leak
 * into reservations. `reserveAll` is atomic: every line is reserved, or none.
 */
export class InMemoryInventoryGateway implements InventoryGateway {
  private readonly stock: Map<Sku, number>;

  private readonly reservations = new Map<string, { sku: Sku; amount: number }[]>();

  /**
   * When true, `release` returns `CompensationFailure` without restoring stock.
   */
  failRelease = false;

  /**
   * Starts from an optional initial stock map, copied defensively.
   *
   * @param stock - initial per-SKU units; defaults to an empty warehouse.
   */
  constructor(stock: ReadonlyMap<Sku, number> = new Map()) {
    this.stock = new Map(stock);
  }

  /**
   * Returns a copy of remaining units per SKU.
   */
  snapshotStock(): Map<Sku, number> {
    return new Map(this.stock);
  }

  /**
   * Reserves every line atomically, or none.
   *
   * @param orderId - the order the reservation belongs to.
   * @param lines - the lines to reserve.
   * @param idempotencyKey - key that replays an existing reservation.
   * @returns a reservation token, or `InsufficientStock` when any line falls short.
   */
  reserveAll(
    orderId: OrderId,
    lines: readonly OrderLine[],
    idempotencyKey: string,
  ): ReservationToken | InsufficientStock {
    if (this.reservations.has(idempotencyKey)) {
      return { orderId, idempotencyKey };
    }
    const needed: { sku: Sku; amount: number }[] = [];
    for (const line of lines) {
      const available = this.stock.get(line.sku) ?? 0;
      if (available < line.quantity.value) {
        return new InsufficientStock(line.sku, line.quantity, available);
      }
      needed.push({ sku: line.sku, amount: line.quantity.value });
    }
    for (const item of needed) {
      this.stock.set(item.sku, (this.stock.get(item.sku) ?? 0) - item.amount);
    }
    this.reservations.set(idempotencyKey, needed);
    return { orderId, idempotencyKey };
  }

  /**
   * Puts reserved units back.
   *
   * @param token - token from a successful `reserveAll`.
   * @returns `undefined` on success, or `CompensationFailure` when configured to fail.
   */
  release(token: ReservationToken): CompensationFailure | undefined {
    if (this.failRelease) {
      return new CompensationFailure("release", "forced failure");
    }
    const held = this.reservations.get(token.idempotencyKey);
    this.reservations.delete(token.idempotencyKey);
    if (held === undefined) {
      return undefined;
    }
    for (const item of held) {
      this.stock.set(item.sku, (this.stock.get(item.sku) ?? 0) + item.amount);
    }
    return undefined;
  }
}

File: python/src/warehouse/adapters/inventory.py
"""In-memory inventory adapter with atomic reserve-all."""

from __future__ import annotations

import threading
from typing import TYPE_CHECKING

from warehouse.application.ports import ReservationToken
from warehouse.domain.errors import CompensationFailure, InsufficientStock

if TYPE_CHECKING:
    from collections.abc import Sequence

    from warehouse.domain.order import OrderId, OrderLine
    from warehouse.domain.sku import Sku


class InMemoryInventoryGateway:
    """InventoryGateway double enforcing finite stock, for tests and demos."""

    def __init__(self, stock: dict[Sku, int] | None = None) -> None:
        """Start from an optional initial stock map, copied defensively."""
        self._lock = threading.Lock()
        self._stock: dict[Sku, int] = dict(stock) if stock is not None else {}
        self._reservations: dict[str, list[tuple[Sku, int]]] = {}
        self.fail_release = False

    def snapshot_stock(self) -> dict[Sku, int]:
        """Return a copy of remaining units per SKU."""
        with self._lock:
            return dict(self._stock)

    def reserve_all(
        self,
        order_id: OrderId,
        lines: Sequence[OrderLine],
        idempotency_key: str,
    ) -> ReservationToken | InsufficientStock:
        """Reserve every line atomically, or none."""
        with self._lock:
            if idempotency_key in self._reservations:
                return ReservationToken(order_id=order_id, idempotency_key=idempotency_key)
            needed: list[tuple[Sku, int]] = []
            for line in lines:
                available = self._stock.get(line.sku, 0)
                if available < line.quantity.value:
                    return InsufficientStock(
                        sku=line.sku,
                        requested=line.quantity,
                        available=available,
                    )
                needed.append((line.sku, line.quantity.value))
            for sku, amount in needed:
                self._stock[sku] = self._stock[sku] - amount
            self._reservations[idempotency_key] = needed
            return ReservationToken(order_id=order_id, idempotency_key=idempotency_key)

    def release(
        self,
        token: ReservationToken,
    ) -> None | CompensationFailure:
        """Put reserved units back."""
        with self._lock:
            if self.fail_release:
                return CompensationFailure(stage="release", detail="forced failure")
            held = self._reservations.pop(token.idempotency_key, None)
            if held is None:
                return None
            for sku, amount in held:
                self._stock[sku] = self._stock.get(sku, 0) + amount
            return None
